### ISSUE TITLE
stdin.getch or read one character

### ISSUE THREAD

**Issue by matti** [socketry/async-io#18]
How to read just one character? Like with normal io:

```
require "io/console"
while c = $stdin.getch do
  p c
end
```

**Comment by ioquatix** [comment]
`io.read(1)`

**Comment by matti** [comment]
```
          Async::Reactor.run do |task|
            async_stdin = Async::IO::Stream.new(
              Async::IO::Generic.new($stdin)
            )
            #async_stdin = Async::IO::Generic.new($stdin)

            while c = async_stdin.read(1) do
              p c
            end
          end
````

^ doesn't work

**Comment by ioquatix** [comment]
Does that example help?

**Comment by matti** [comment]
yes, thanks! I don't fully understand why I need to set `$stdin.raw = true` because with my original example code I don't have to.

**Comment by matti** [comment]
(see here: https://github.com/matti/runcmd-cli/blob/master/lib/runcmd/cli/record_command.rb)

**Comment by ioquatix** [comment]
Ruby 'getch' does it for you automatically.

**Comment by ioquatix** [comment]
IMHO no.

Ruby needs a better terminal library for dealing with this stuff. There is more involved than just reading characters from the input stream. Combining it all into IO was a mistake, and why (some of) it lives in `io/console`.

**Comment by matti** [comment]
Totally agree with that. But from my mediocre developer POV who is extremely excited about this async-* stuff it would follow the "principle of least surprise" that got me excited about ruby in 2005.

**Comment by ioquatix** [comment]
Ideally we'd wrap the method directly on the IO object but due to the weird hybrid design of non-blocking IO on Ruby, it's simply not possible (i.e. there is no `getch_nonblock`). This is going in the too hard box for now.

**Comment by ioquatix** [comment]
Here is a PR which makes all IO non-blocking so naturally you could just use `$stdin.getch`: https://github.com/ruby/ruby/pull/1870

**Comment by phobetron** [comment]
Hoping that PR gets in a release soon 🤞 `$stdin.raw!` jacks up all my output, and I haven't figured out how to work around that.

**Comment by matti** [comment]
That PR has been closed because it was not updated

**Comment by phobetron** [comment]
It was closed due to inactivity with conflicts. It can be reopened when they're resolved.

**Comment by ioquatix** [comment]
I’m working on it but it’s not easy.
